README.md
# ejs-template-server

Ever wanted to automatically generate OpenGraph images based on a set of input parameters? This package makes it possible to spin up a short-lived server that can tranform a given template on request based un URL query parameters using a simple API.

## Installation

For most use cases, installing this package as a devDependency should be enough.

### Yarn

```bash
yarn add -D ejs-template-server
```

### NPM

```bash
npm install -D ejs-template-server
```

## Usage

### Simple

Add HTML with EJS syntax to the `raw` config option. When the server is created, EJS will parse the input for required variables, and will subsequently accept requests that contain these variables in its query params.

**`server.ts`**

```ts
import { createEjsTemplateServer } from 'ejs-template-server'
import { URLSearchParams } from 'url'
import axios from 'axios'

createEjsTemplateServer({
  template: {
    path: '/og-template.html',
  },
})
  .then(() => {
    const params = new URLSearchParams({
      title: 'Hello world!',
    }).toString()

    const result = axios.get(`http://localhost:8080?${params}`)

    // Do what you want with the result, see the advanced section for an example
  })
  .catch((e) => {
    console.log('The server was unable to start.', e)
    exit(1)
  })
```

### Advanced

Add the path to a file that you want to use as a template to the `path` config option. Like in the simple example, EJS will automatically recognize required parameters and check the query parameters for these.

In this example, however, [`playwright`](https://playwright.dev/) is used in order to programatically open the page and take a screenshot at a certain size.

**`server.ts`**

```tsx
import fs from 'fs'
import { createEjsTemplateServer } from 'ejs-template-server'
import { fileURLToPath, URLSearchParams } from 'url'
import { launchChromium } from 'playwright-aws-lambda'

const PORT = 8081

createEjsTemplateServer({
  template: {
    path: 'og-template.html',
  },
  staticParams: {
    getFontSize: (title: string) => (title?.length > 32 ? `text-7xl` : `text-9xl`),
  },
  port: PORT,
})
  .then(() => {
    generateImage('http://localhost:' + PORT)
  })
  .catch((e) => {
    console.log('The server was unable to start.', e)
    exit(1)
  })

function generateImage(url: string) {
  const params = new URLSearchParams({
    title: 'Hello world!',
    date: '2021-11-20',
    readTime: '2 minutes',
  }).toString()

  const url = `${baseURL}?${new URLSearchParams(params).toString()}`

  try {
    const browser = await launchChromium({ headless: true })
    const page = await browser.newPage()
    await page.setViewportSize({ width: 1200, height: 630 })
    await page.goto(url, { waitUntil: 'networkidle' })

    const buffer = await page.screenshot({ type: 'png' })
    await browser.close()

    fs.writeFileSync('output.png', buffer)
  } catch (error) {
    console.error(`An error occurred while generating the image.`, error)
    exit(1)
  }
}
```

**`og-template.html`**

```html
<main>
  <p>
    <% if (date) { %>
    <span><%= date %></span>
    <% } %> — <% if (readTime) { %>
    <span><%= readTime %></span>
    <% } %>
  </p>
  <h1 class="<%= getFontSize(title) %>"><%= title %></h1>
</main>
```

## Options

The `createEjsTemplateServer` function only _requires_ that you provide a template (path or raw), but there are a few other options as well.

| Option       | Description                                                                     | Required | Type                                                     | Default |
| ------------ | ------------------------------------------------------------------------------- | -------- | -------------------------------------------------------- | ------- |
| template     | An object containing either a `raw` or a `path` key respectively.               | yes      | { raw?: string, path?: string }                          |         |
| port         | The port that the server should listen on.                                      | no       | 8080                                                     | 8080    |
| staticParams | Non-changing parameters that should be available at all times in the EJS scope. | no       | Record<string, string \| ((params: unknown) => unknown)> |         |

## Contributing

For guidelines and useful information, please see [CONTRIBUTING.md](https://github.com/SkySails/ejs-template-server/blob/main/CONTRIBUTING.md)

## License

[MIT](https://github.com/SkySails/ejs-template-server/blob/main/LICENSE)


CHANGELOG.md
## [1.0.3](https://github.com/SkySails/ejs-template-server/compare/v1.0.2...v1.0.3) (2021-11-24)


### Bug Fixes

* allow any parameters in staticParams function ([67e9016](https://github.com/SkySails/ejs-template-server/commit/67e90167340006f8cbf218a81b559a1a0d6b2d13))

## [1.0.2](https://github.com/SkySails/ejs-template-server/compare/v1.0.1...v1.0.2) (2021-11-23)


### Bug Fixes

* add real useful information to readme ([1c6397b](https://github.com/SkySails/ejs-template-server/commit/1c6397bf7f323ada7fa6caed16408b2dcbd8f702))

## [1.0.1](v1.0.0...v1.0.1) (2021-11-22)

### Bug Fixes

- add extension to import for mjs compat ([d681d92](d681d92))

# 1.0.0 (2021-11-22)

### Features

- add `createEjsTemplateServer` function ([cedb818](cedb818))
- make it possible to force-quit server ([d043277](d043277))


test/index.e2e-spec.ts
import { createEjsTemplateServer, EjsTemplateConfig } from '../src/index'
import puppeteer, { Browser, Page } from 'puppeteer'
import { Server } from 'http'

const createServer = async (template: EjsTemplateConfig) =>
  createEjsTemplateServer({
    template,
    port: 8456,
  })

describe('EJS Template Server', () => {
  let browser: Browser
  let page: Page
  let server: Server

  beforeAll(async () => {
    browser = await puppeteer.launch()
    page = await browser.newPage()
    jest.spyOn(console, 'error').mockImplementation(() => {})
  })

  it("given the template '<h1><%= title %><h1>', when the query is 'Hello world!' should render a heading tag with the text 'Hello world'", async () => {
    server = await createServer({ raw: '<h1><%= title %></h1>' })
    await page.goto('http://localhost:8456?title=Hello%20world%21')
    const heading = await page.waitForSelector('h1')
    const text = await page.evaluate((el) => el.textContent, heading)
    expect(text).toContain('Hello')
  })

  it("given the template '<h1><%= title %><h1>', when the query is empty, should log error with missing params & have status 403", async () => {
    server = await createServer({ raw: '<h1><%= title %></h1>' })
    const response = await page.goto('http://localhost:8456')
    expect(response.status()).toBe(403)
    expect(console.error).toHaveBeenCalledWith('Received invalid request:', ['title is not defined'])
  })

  it("given a path to a test template, when the query is 'Hello world', should render a heading tag with the text 'Hello world' ", async () => {
    server = await createServer({ path: 'test/test-template.html' })
    await page.goto('http://localhost:8456?title=Hello%20world%21')
    const heading = await page.waitForSelector('h2')
    const text = await page.evaluate((el) => el.textContent, heading)
    expect(text).toContain('Hello')
  })

  it('given a path to an non-existing template, should throw', async () => {
    expect(() => createServer({ path: 'nonexistent' })).rejects.toThrowError(
      'ENOENT: Invalid path provided, no template was found',
    )
  })

  afterAll(() => {
    browser.close()
  })

  afterEach(async () => {
    jest.resetAllMocks()

    await new Promise((r) => {
      server.close(r)
    })
  })
})


test/jest-e2e.config.js
const reporters = ["default"];

if (process.env.NODE_ENV.toUpperCase() === "CI")
  reporters.push([
    "jest-junit",
    {
      outputName: "e2e.xml",
    },
  ]);

/** @type {import('@jest/types').Config.InitialOptions} */
const config = {
  moduleFileExtensions: ["js", "json", "ts"],
  rootDir: ".",
  testEnvironment: "node",
  testRegex: ".e2e-spec.ts$",
  transform: {
    "^.+\\.(t|j)s$": "ts-jest",
  },
  reporters,
};

module.exports = config;


src/index.ts
import ejs from 'ejs'
import { existsSync } from 'fs'
import http from 'http'
import { resolve } from 'path'
import URL from 'url'
import { createCloseableHttpServer } from './lib/httpserver.js'
import type { XOR } from './types'

export type EjsTemplateConfig = XOR<{ raw: string }, { path: string }>

export interface EjsTemplateServerOptions {
  /** The (EJS compatible) HTML template to use when rendering the OG image.
   * You must specify *only one* of `raw` and `path`.
   * @see https://ejs.co/
   */
  template: EjsTemplateConfig
  /** The port that the server will try to listen on.
   * @default 8080
   */
  port?: number
  staticParams?: Record<string, string | ((params: any) => unknown)>
}

/**
 * Creates a server that is capable of rendering OG images on demand.
 * @param  {EjsTemplateServerOptions} options
 * @async
 * @returns {Promise<http.Server>}
 */
export async function createEjsTemplateServer({
  template,
  port = 8080,
  staticParams,
}: EjsTemplateServerOptions): Promise<http.Server> {
  if (!template.raw && !template.path) {
    throw new Error('No template provided, cannot start OpenGraph server.')
  }

  if (template.path && !existsSync(resolve(template.path))) {
    throw new Error('ENOENT: Invalid path provided, no template was found')
  }

  const handleError = (e: Error, res: http.ServerResponse) => {
    const undefinedError = String(e).match(/.*is not defined/g)
    if (undefinedError) {
      console.error('Received invalid request:', undefinedError)
      res.statusCode = 403
      res.setHeader('Content-Type', 'application/json').end(JSON.stringify(undefinedError))
    } else {
      res.statusCode = 500
      console.error('An error occurred while trying to respond to a request:', e)
      res.setHeader('Content-Type', 'application/json').end(JSON.stringify(e))
    }
  }

  const srv = createCloseableHttpServer(async (req, res) => {
    if (req.url) {
      if (template.path) {
        try {
          const html = await ejs.renderFile(template.path, { ...URL.parse(req.url, true).query, ...staticParams }); // prettier-ignore

          return res.end(html)
        } catch (e: any) {
          handleError(e, res)
        }
      } else if (template.raw) {
        try {
          const html = ejs.render(template.raw, { ...URL.parse(req.url, true).query, ...staticParams }); // prettier-ignore
          return res.end(html)
        } catch (e: any) {
          handleError(e, res)
        }
      }
    } else {
      return res.end('Invalid url.')
    }
  }).listen(port)

  return Promise.resolve(srv)
}
